src/nwbuild.js

```diff
@@ -1,96 +1,148 @@
+    } else {
+      throw new Error(
+        `nwbuild property in the ${
+          options.srcDir
+        }/package.json is of type ${typeof nwPkg.nwbuild}. Expected type object.`,
+      );
+    }
 
-  try {
-    await access(nwDir, constants.F_OK);
-  } catch (e) {
-    fileExists = false;
-  }
+    // Parse options, set required values to undefined and flags with default values unless specified by user
+    options = await parse(options);
 
-  if (noCache === true || fileExists === false) {
-    await rm(nwDir, { force: true, recursive: true });
-    await download(version, flavour, platform, arch, downloadUrl, cacheDir);
-    await decompress(platform, cacheDir);
-    await remove(platform, cacheDir);
-  }
+    // Create cacheDir if it does not exist
+    await mkdir(options.cacheDir, { recursive: false });
+
+    // Validate options.version here
+    // We need to do this to get the version specific release info
+    releaseInfo = await getReleaseInfo(
+      options.version,
+      options.cacheDir,
+      options.manifestUrl,
+    );
+
+    // TODO: validate options
+    const e = await validate(options, releaseInfo);
+    if (e === false) {
+      throw new Error("Some option was invalid.");
+    }
+
+    // Get current platform and arch if mode is run
+    if (options.mode === "run") {
+      let tmpPlatform = await getPlatform(platform);
+      let tmpArch = await getArch(arch);
+      if (tmpPlatform === undefined) {
+        throw new Error(`Platform ${platform} is not supported. Sorry!`);
+      }
+      if (tmpArch === undefined) {
+        throw new Error(`Architecture ${arch} is not supported. Sorry!`);
+      } else {
+        options.platform = tmpPlatform;
+        options.arch = tmpArch;
+      }
+    }
 
-  let releaseInfo = await getReleaseInfo(version, cacheDir, manifestUrl);
+    // Variable to store nwDir file path
+    nwDir = `${options.cacheDir}/nwjs${
+      options.flavour === "sdk" ? "-sdk" : ""
+    }-v${options.version}-${options.platform}-${options.arch}`;
 
-  if (run === true) {
-    await develop(srcDir, nwDir, platform);
-  } else {
-    await packager(srcDir, nwDir, outDir, platform, zip, releaseInfo);
+    // Download relevant NW.js binaries
+    let cached = await isCached(nwDir);
+    if (options?.noCache === true || cached === false) {
+      await rm(nwDir, { force: true, recursive: true });
+      await download(
+        options.version,
+        options.flavour,
+        options.platform,
+        options.arch,
+        options.downloadUrl,
+        options.cacheDir,
+      );
+      await decompress(options.platform, options.cacheDir);
+      await remove(options.platform, options.cacheDir);
+    }
+
+    if (options.mode !== "run" && options.mode !== "build") {
+      throw new Error("Invalid mode value. Expected run or build.");
+    }
+    if (options.mode === "run") {
+      await develop(options.srcDir, nwDir, options.platform);
```

Comment:
Since the only options are "run" or "build", you should probably `return` after this `develop()` call. No need to run through the rest of the function.

As an additional change, you could even remove the `if (options.mode === "build")` (after adding the `return` above). This would make "build" the default mode, so you could also remove the check on lines 128-130.

Reply:
So, I haven't had a chance to review the changed code yet, but I see the new documentation still doesn't have defaults for `mode`, `version`, or `flavour` (note that you should probably support either `flavor` or `flavour`). I think all of these properties should have defaults:
`mode`: default to "build"
`version`: default to "latest"
`flavor`: default to "normal"

Can probably even get away with defaults for `platform` (default to either the current OS or every platform), `arch` (default to "x64"), and `outDir` (default to "./dist/").

Reply:
Is it best practice for API properties to have defaults?

The reason I didn'twant these arguments to have default values is because the "default" is derived from the `https://nwjs.io/versions` manifest file. A feature of nw-builder is the ability to define custom manifest files with their own defaults. That's why I thought it would not makes sense to have default values for only one type of manifest.

Reply:
> **NW.js Arch default:** 32-bit on Windows, 64-bit on Linux/OSX

What?? 64-bit everywhere...

Reply:
> Yes it is better to have defaults for everything, so as the API grows, your usage of it only requires overriding the parts you need to customize. Also as users upgrade the library to newer versions, presumably new API options will be added, so they shouldn't have to update all their code every time to add in default values for new properties.

This makes so much sense!

> **NW.js Arch default"**: 32-bit on Windows, 64-bit on Linux/OSX

Currently it gets the default architecture from `process.arch` so I don't think hardcoding a default arch is necessary.

Reply:
Good to know!